### DISCUSSION TITLE
ANT+ with raspberrypi

### DISCUSSION THREAD

**Discussion by chreds** [cagnulein/qdomyos-zwift#1685] [Q&A] [answered]
I just got a raspberry pi zero w and followed the installation instructions. All is good except I can't get the device to read heart rate from my garmin watch.

I have my watch set to broadcast ANT+ heartrate and I have an ANT+ usb dongle plugged into the pi. I'm guessing there's some config or ANT libraries I need to install for this to work? Is this documented anywhere? I couldn't find anything.

Also-- how can I set config options that aren't available with the existing args? I'd like to set the echelon gears control knob on.

I'm using pi zero w headless with -no-gui.

Thanks!

**Comment by cagnulein** [comment, marked as answer]
hi @chreds unfortunately i never started the work to port the ant+ on raspberry. ant+ is supported only on android. i suggest to use your phone instead

**Comment by chreds** [comment]
No worries. I will just put the ant stick on the tablet and connect the hr monitor directly to rouvy.

What about config settings when using no-gui? How can I turn in echelon gear changing via the knob for headless raspberry operation?

**Comment by cagnulein** [comment]
you can edit the settings from the config file under .config folder or you
can create a pull request adding a new --setting for the gears

Il giorno lun 2 ott 2023 alle 06:47 chreds ***@***.***> ha
scritto:

**Reply by chreds** [reply]
Thanks. I really appreciate your work on this project! I've bought the android app from both stores now and will probably toss some more your way here.

**Comment by cagnulein** [comment]
thanks! let me know if you need help

Il giorno lun 2 ott 2023 alle 08:02 chreds ***@***.***> ha
scritto:

**Reply by chreds** [reply]
Have you ever tried the webgl qt platform plugin? Could be interesting way to connect to the app without needing a vnc client.

**Reply by cagnulein** [reply]
nope i never tried @chreds 
give it a try and let me know in case :)

**Reply by chreds** [reply]
Works pretty good. Definitely a bit laggy on the raspberry pi zero w, but maybe with a higher powered pi, or if I increase the GPU ram allocation it will be better.

I like this since I can load it up in browser on phone but the raspberry is doing all the connections to the bike so it can keep the same bluetooth id, vs android that couldn't.

![image](https://github.com/cagnulein/qdomyos-zwift/assets/7526036/81488c2b-b68a-4be8-8fa2-c080973ff2b3)

```
sudo apt install qtbase5-private-dev

cd ~
wget https://download.qt.io/official_releases/qt/5.15/5.15.2/submodules/qtwebglplugin-everywhere-src-5.15.2.zip
 
unzip qtwebglplugin-everywhere-src-5.15.2.zip
cd qtwebglplugin-everywhere-src-5.15.2

qmake
make

sudo cp plugins/platforms/libqwebgl.so /usr/lib/arm-linux-gnueabihf/qt5/plugins/platforms/

cd qdomyos-zwift/src
sudo ./qdomyos-zwift -qml -platform webgl:port=80
```

Default port is 8080 if you don't set it.

Then launch browser to the IP address of the pi.

**Reply by chreds** [reply]
If anyone else is looking for the config file, it's at /root/.config/Roberto Viola/qDomyos-Zwift.conf if you are using the raspberry pi install instructions here (https://github.com/cagnulein/qdomyos-zwift/blob/master/docs/10_Installation.md) with systemd running the service.

Also performance is better with GPU RAM at 128 in raspi-config > Performance but obviously this will slow down compiling so maybe don't change that before compiling it all.

**Reply by cagnulein** [reply]
wow looks amazing! thanks for reporting it! maybe we can create a wiki page for it, what do you think?

**Reply by chreds** [reply]
Yeah feel free to use this info however you want. It could be useful to include a config section after the raspberry pi installation instructions that provides things like where the config file is, links to explanations of those options (I think i saw this on your personal domain).

Editing settings is painful from the phone via webgl, but the main UI is fine. Probably would want to set zoom to like 80% or something so 2 or more tiles fit side by side on the phone. The slide out animation for settings is really laggy. I wonder if performance would be better on a zero 2 w? Hard to find those right now.

**Reply by dstrimble** [reply]
This sure doesn't work for me. When I add -qml -platform webgl:port=80 to my systemctl command file it just causes the app to fail to startup  (code=exited, status=255/EXCEPTION)

**Reply by cagnulein** [reply]
@dstrimble did you install the webgl module? I never tried this method so the only person that can help here is @chreds

**Reply by dstrimble** [reply]
I followed the steps he posted exactly, yes.

**Reply by cagnulein** [reply]
@dstrimble I don't know, maybe some other services on the raspberry with the same port? You can try to run gdb and check where the issue starts

**Reply by chreds** [reply]
My first guess would be the port's already in use. Does syslog have anymore information on the error? You can try netstat -nl to see if anything is listening on 80 already, or try changing it to a different port and starting. 

If the QT version gets updated in QZ I imagine that would require you to update the plugin version as well but I doubt that's happened in the last few weeks.

Could be some packages missing too. I did the above commands after doing the install here: https://github.com/cagnulein/qdomyos-zwift/blob/master/docs/10_Installation.md

**Reply by dstrimble** [reply]
Nothing spectacular in syslog about it. I even stopped it with sysctl and tried starting it directly with sudo ./qdomyos-zwift -qml -platform webgl:port=80 and I just get this:
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
OK, you are root.
and it immediately goes back to command prompt.
netstat shows nothing using 80, even after I run the command.

**Reply by cagnulein** [reply]
@dstrimble  does it work without webgl?

**Reply by dstrimble** [reply]
Yes, QZ works fine without the webgl parameters, but I don't have a UI to speak of.

**Reply by chreds** [reply]
What hardware are you on?

**Comment by cagnulein** [comment]
yeah it could be  nice but i don't have time to also maintain the raspberry
branch. i'm already receiving hundreds of email every day for ios and
android :)

i found the raspi 4 "enough" for the UI :)


Il giorno lun 2 ott 2023 alle 20:27 chreds ***@***.***> ha
scritto:

**Reply by chreds** [reply]
And this is a hobby project for you? Oi.

How much of the code base is shared?

**Comment by cagnulein** [comment]
yes, it's only an hobby. i have a daily job as a firmware developer.

the code here is the 100%

Il giorno lun 2 ott 2023 alle 20:38 chreds ***@***.***> ha
scritto:

**Reply by chreds** [reply]
I mean how much of the code base is shared between the 3 platforms?

**Comment by cagnulein** [comment]
hah sorry, let me say 95%

Il giorno lun 2 ott 2023 alle 20:43 chreds ***@***.***> ha
scritto:

**Reply by chreds** [reply]
I don't mind helping contribute this info to the wiki. Where would you want it?

Here?
https://github.com/cagnulein/qdomyos-zwift/wiki

**Comment by cagnulein** [comment]
yeah sure create a new page here
https://github.com/cagnulein/qdomyos-zwift/wiki/_new

thank you very much !

Il giorno lun 2 ott 2023 alle 20:46 chreds ***@***.***> ha
scritto:

**Comment by cagnulein** [comment]
which version of qt are you using?

Il giorno mer 25 ott 2023 alle 17:48 Daniel Trimble <
***@***.***> ha scritto:

**Reply by dstrimble** [reply]
I downloaded from git and ran the make as outlined in these instructions, so it should be latest. https://github.com/cagnulein/qdomyos-zwift/blob/master/docs/10_Installation.md#qdomyos-zwift-installation

**Comment by cagnulein** [comment]
please check just to be sure. it could be the issue

Il giorno mer 25 ott 2023 alle 17:56 Daniel Trimble <
***@***.***> ha scritto:

**Reply by dstrimble** [reply]
Where do I get the version?

**Comment by cagnulein** [comment]
qmake --version


Il giorno mer 25 ott 2023 alle 18:14 Daniel Trimble <
***@***.***> ha scritto:

**Reply by dstrimble** [reply]
QMake version 3.1
Using Qt version 5.15.2 in /usr/lib/arm-linux-gnueabihf

**Comment by cagnulein** [comment]
right so.
the only way to understand something more is to run qz with gdb

so like

gdb --args ./qdomyos-zwift .......

and try to catch the back trace

Il giorno mer 25 ott 2023 alle 18:29 Daniel Trimble <
***@***.***> ha scritto:

**Reply by dstrimble** [reply]
Starting program: /home/user/qdomyos-zwift/src/qdomyos-zwift -qml -platform webgl:port=80
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/arm-linux-gnueabihf/libthread_db.so.1".
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
[New Thread 0x71778380 (LWP 3272)]
OK, you are root.
[New Thread 0x70dff380 (LWP 3273)]
[New Thread 0x703ff380 (LWP 3274)]
[Detaching after fork from child process 3465]
[New Thread 0x6d8db380 (LWP 3466)]
[Detaching after fork from child process 3467]
[Thread 0x703ff380 (LWP 3274) exited]
[Thread 0x6d8db380 (LWP 3466) exited]
[Thread 0x71778380 (LWP 3272) exited]
[Thread 0x70dff380 (LWP 3273) exited]
[Inferior 1 (process 3267) exited with code 0377]

**Comment by cagnulein** [comment]
are you able to type

bt

to collect the backtrace?

Il giorno mer 25 ott 2023 alle 18:40 Daniel Trimble <
***@***.***> ha scritto:

**Reply by dstrimble** [reply]
(gdb) bt
No stack.

**Comment by cagnulein** [comment]
maybe a ram issue? raspberry model?

Il giorno mer 25 ott 2023 alle 18:43 Daniel Trimble <
***@***.***> ha scritto:

**Reply by dstrimble** [reply]
Raspberry Pi 3 Model B Rev 1.2

**Comment by cagnulein** [comment]
dmesg says something after the crash?

Il giorno mer 25 ott 2023 alle 18:46 Daniel Trimble <
***@***.***> ha scritto:

**Reply by dstrimble** [reply]
dmesg dumps a lot of info. Do you want it all?

**Comment by cagnulein** [comment]
dmesg > log.txt and attach here

maybe we will something useful

Il giorno mer 25 ott 2023 alle 19:00 Daniel Trimble <
***@***.***> ha scritto:

**Reply by dstrimble** [reply]
[log.txt](https://github.com/cagnulein/qdomyos-zwift/files/13169205/log.txt)

**Reply by cagnulein** [reply]
@chreds maybe can you share the full image of the sdcard?

**Comment by chreds** [comment]
In raspi-config:
Performance Options-> How much memory do you have allocated to the GPU?
Advanced Options-> Which GL driver?
I'm at 128 / legacy

**Reply by dstrimble** [reply]
64/legacy

I can try 128

**Reply by dstrimble** [reply]
No luck, same thing with 128

**Comment by chreds** [comment]
@cagnulein how do you want the image? tgz of the sd card or clonezilla or gzip'd dd? Do you want both boot partition and data?

**Comment by cagnulein** [comment]
tgz of the whole sdcard will be perfect, thanks!

Il giorno gio 26 ott 2023 alle 19:16 chreds ***@***.***> ha
scritto:

**Comment by chreds** [comment]
Here it is (1.65GB) 
https://1drv.ms/u/s!Ajw0OnFQgXJ5iy87eGNMn42P2ZSd?e=mMhXhC

Let me know when you've got it downloaded and I'll delete it.

**Comment by cagnulein** [comment]
Awesome, thanks!

Il giorno mer 1 nov 2023 alle 17:21 chreds ***@***.***> ha
scritto:

**Reply by chreds** [reply]
Was that of any help? Did you grab it yet can I remove it?

**Reply by cagnulein** [reply]
@dstrimble did you get it?

**Reply by cagnulein** [reply]
@chreds i downaloaded it for him if you would like to remove it to free the space on your onedrive.

**Reply by cagnulein** [reply]
@chreds @dstrimble i uploaded the image also on my drive if you want to remove from yours https://drive.google.com/file/d/1Zk27ArTUFb-2ySLpSSZMSHpLpxI1n1y9/view?usp=sharing
